### DISCUSSION TITLE
Issues getting started

### DISCUSSION THREAD

**Discussion by isaaccloos** [metalbear-co/mirrord#135] [Q&A] [answered]
Coming from [TWIR 446](https://this-week-in-rust.org/blog/2022/06/08/this-week-in-rust-446/) looking to tackle one of the "Good First Issue"(s). I ran into complications getting the development environment up and running, or even `cargo check`(ing) for that matter.

```
root ➜ /workspaces/mirrord (main) $ cargo check
warning: /workspaces/mirrord/tests/Cargo.toml: unused manifest key: dependencies.mirrord.bin
   Compiling frida-gum-sys v0.3.4
error: failed to run custom build command for `frida-gum-sys v0.3.4`

Caused by:
  process didn't exit successfully: `/workspaces/mirrord/target/debug/build/frida-gum-sys-525cada622f8b263/build-script-build` (exit status: 101)
  --- stdout
  cargo:rustc-link-search=/usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/frida-gum-sys-0.3.4
  cargo:rustc-link-search=/workspaces/mirrord/target/debug/build/frida-gum-sys-7148217ad4b77ab6/out
  cargo:rustc-link-lib=static=frida-gum
  cargo:rustc-link-lib=pthread

  --- stderr
  thread 'main' panicked at 'Unable to find libclang: "couldn't find any valid shared libraries matching: ['libclang.so', 'libclang-*.so', 'libclang.so.*', 'libclang-*.so.*'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.57.0/src/lib.rs:1975:31
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I have given an effort to resolve this issue, but no luck. I am running an Apple Silicon M1 Pro macbook with rosetta installed. After first seeing this error message I downloaded the libclang.so from https://releases.llvm.org/ and attempted to reference the libclang.so file the error message specified via an environment variable like so `LIBCLANG_PATH={path} cargo check` but that also resulted in the same error. 

Perhaps this has something to do with this project running in a container?

Any advice would be much appreciated, thanks!

**Comment by infiniteregrets** [comment, marked as answer]
Hi! Thanks for reaching out. What is the path you are setting `LIBCLANG_PATH` to? Is it the bin directory of the install? Have you tried installing with homebrew? `brew install --with-toolchain llvm`

> Perhaps this has something to do with this project running in a container?

It is required for FFI bindings and doesn't have to do anything with containers

**Reply by isaaccloos** [reply]
Thanks a lot for the response 😄

Here are the variations I tried originally, before your homebrew tip:
`LIBCLANG_PATH=/Users/{me}/Downloads/clang+llvm-13.0.1-amd64-unknown-freebsd12/lib/ cargo check` (this does have libclang.so file in it, but didn't seem to work)
`LIBCLANG_PATH=/Users/{me}/Downloads/clang+llvm-13.0.1-amd64-unknown-freebsd12/lib/libclang.so cargo check` (direct link)
`LIBCLANG_PATH=/Users/{me}/Downloads/clang+llvm-13.0.1-amd64-unknown-freebsd12/bin/ cargo check` (exec(s) available, but also no luck)

after reading your advice I installed llvm via homebrew:
`brew install llvm` (`--with-toolchain` was an unavailable command on my end. Found something about that [here](https://github.com/pytorch/glow/issues/2389) and [here](https://stackoverflow.com/questions/57788632/homebrew-error-invalid-option-with-clang))

Attempted a few new path targets:
`LIBCLANG_PATH=/opt/homebrew/Cellar/llvm/13.0.1_1/lib/ cargo check`
`LIBCLANG_PATH=/opt/homebrew/Cellar/llvm/13.0.1_1/bin/ cargo check`
`LIBCLANG_PATH=/opt/homebrew/Cellar/llvm/13.0.1_1/bin/Clang-13 cargo check` (I know that's not even what the panic defined, but I was getting desperate)

So far no luck. Gladly accepting additional pointers.

p.s.
I've been documenting my journey through this so far. Once I finally succeed hopefully that information will be helpful to another developer looking to contribute too.

**Reply by aviramha** [reply]
Sorry for the troubles @IsaacCloos and we'd love having docs for newcomers (I was about to create a "compiling from source docs" issue so we'll do it later!) If you can do it it'd be amazing.

Regarding your current troubles, you're having hard time compiling [frida](frida.re) which is a dependency we have.
Are you running inside a container or directly on your machine? and if on a container, is it arm or x86?

**Reply by aviramha** [reply]
Can you check if the llvm you installed is x64/arm?

**Reply by isaaccloos** [reply]
Sorry for delay! I was pretty determined to get this completely ironed out before reporting back. However I have yet to cross the finish line. You are absolutely correct that `frida-gum` is causing a slew of complications preventing me from building (or `check`ing) `mirrord-layer`, and thus `mirrord-cli` (which has layer as a dependency). I have been able to successfully check `mirrord-agent` and `mirrord-protocol` individually. 

I am running inside of the container that the `main` branch of the repository prompts to deploy and mount via the `Remote - Containers` vscode extension.

The container is running arm64 architecture; Debian “bullseye” 
```
root ➜ /workspaces/mirrord (main ✗) $ arch
aarch64
```

I was able to get past the llvm issue by running the following codes in the containers environment:
```
apt-get install llvm
apt-get install clang
```

I also needed to install rustfmt:
```
rustup component add rustfmt --toolchain nightly-aarch64-unknown-linux-gnu
```

Could these extra steps be defined by the image? There's also a chance I didn't initialize the container correctly, if that's the case sorry for the trouble!

## Current blocker

```
    Checking frida-gum-sys v0.3.4
error[E0432]: unresolved import `_frida_g_object_unref`
  --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/frida-gum-sys-0.3.4/src/lib.rs:25:9
   |
25 | pub use _frida_g_object_unref as g_object_unref;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no external crate `_frida_g_object_unref`

For more information about this error, try `rustc --explain E0432`.
The following warnings were emitted during compilation:

warning: Frida gum devkit not found, downloading from https://github.com/frida/frida/releases/download/15.1.17/frida-gum-devkit-15.1.17-linux-arm64.tar.xz...

error: could not compile `frida-gum-sys` due to previous error
```

First off, that message about the gum devkit downloading does not resolve the issue. I haven't been able to verify if the download is even successful. Either way the issue persists.

I have been reading through the [frida-rust](https://github.dev/frida/frida-rust) source code, readme(s), and documentation. I [found](https://crates.io/crates/frida-gum) that you can manually download and "use" the devkit. I am currently in the process of getting that to work. So far no luck, but I am new to using .c/.h/.a files with Rust.

If you have advice regarding this particular issue please let me know 😄  

> These posts have been deliberately long winded in the hope that they are more helpful! If you were serious about wanting a guide I'd be happy to contribute something more concise.

**Reply by aviramha** [reply]
> Could these extra steps be defined by the image? There's also a chance I didn't initialize the container correctly, if that's the case sorry for the trouble!

Yeah sure, that'd be great.
Can you try a cargo clean + cargo build?

**Reply by isaaccloos** [reply]
Success! Unfortunately, I had to alter some of the source code for [frida-rust](https://github.com/frida/frida-rust) to get everything working. Given this outcome I am not comfortable recommending similar steps for anyone else to replicate. 

In my testing so far everything seems to be working as intended [mirrord](https://github.com/metalbear-co/mirrord)-side 👍🏻 . For an ultimate solution I believe one of the following things needs to happen:
1. frida-rust improves installation instructions
2. frida-rust code issues are addressed / expounded upon to improve compatibility transparency 

Anywho! I will tentatively call this resolved as I can now `build`, `check`, and `run` this project from the development container. 

Thanks so much for your help 😄

**Reply by aviramha** [reply]
> Success! Unfortunately, I had to alter some of the source code for [frida-rust](https://github.com/frida/frida-rust) to get everything working. Given this outcome I am not comfortable recommending similar steps for anyone else to replicate.

Can you share how you solved it please?

**Reply by isaaccloos** [reply]
> Can you share how you solved it please?

Oh dear! I just noticed this question and what's worse you've caught me without my personal laptop (where I keep my notes) 😔 

I will follow up tonight with all my documentation 📓
